examples/rust/udp-ping-pong.rs

```diff
@@ -185,96 +166,55 @@ impl UdpClient {
         bufsize_bytes: usize,
         num_pings: usize,
     ) -> Result<()> {
-        let mut qtokens: Vec<QToken> = Vec::new();
-
         if let Err(e) = self.libos.bind(self.sockqd, local_socket_addr) {
             anyhow::bail!("bind failed: {:?}", e)
         };
 
-        // Push and pop data.
-        self.sga = Some(mksga(&mut self.libos, bufsize_bytes, fill_char)?);
-        match self.libos.pushto(
-            self.sockqd,
-            &self.sga.expect("should be a valid sgarray"),
-            remote_socket_addr,
-        ) {
-            Ok(qt) => qtokens.push(qt),
-            Err(e) => anyhow::bail!("push failed: {:?}", e),
-        };
-        match self.libos.sgafree(self.sga.expect("should be a valid sgarray")) {
-            Ok(_) => self.sga = None,
-            Err(e) => anyhow::bail!("failed to release scatter-gather array: {:?}", e),
-        }
-        match self.libos.pop(self.sockqd, None) {
-            Ok(qt) => qtokens.push(qt),
-            Err(e) => anyhow::bail!("pop failed: {:?}", e),
-        };
-
-        // Send packets.
-        let mut i: usize = 0;
-        while i < num_pings {
-            let (index, qr): (usize, demi_qresult_t) = match self.libos.wait_any(&qtokens, Some(TIMEOUT_SECONDS)) {
-                Ok((index, qr)) => (index, qr),
-                Err(e) => anyhow::bail!("operation failed: {:?}", e),
+        let mut received_responses: usize = 0;
+        while received_responses < num_pings {
+            let sga: demi_sgarray_t = mksga(&mut self.libos, bufsize_bytes, fill_char)?;
+            // Send packet and wait for response.
+            let returned_sga: demi_sgarray_t = loop {
+                issue_pushto(&mut self.libos, self.sockqd, remote_socket_addr, &sga)?;
+
+                // Wait for the response.
+                let qt: QToken = match self.libos.pop(self.sockqd, None) {
+                    Ok(qt) => qt,
+                    Err(e) => anyhow::bail!("pop failed: {:?}", e),
+                };
+
+                match self.libos.wait(qt, Some(RETRY_TIMEOUT_SECONDS)) {
+                    Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_POP => break unsafe { qr.qr_value.sga },
+                    Ok(_) => anyhow::bail!("unexpected result"),
+                    // Retry if we didn't receive a response in a second.
+                    Err(e) if e.errno == libc::ETIMEDOUT => continue,
```

Comment:
Would you like to track this case by tracing? It might be useful to figure out in debug mode whether this is happening at all and how many times.

Reply:
Sure, will add